plocate-build.8
.TH plocate\-build 8 "Oct 2020" plocate\-build
.SH NAME
plocate\-build \- generate index for plocate

.SH SYNOPSIS
.B plocate-build
.I "[OPTION]..."
.I "MLOCATE_DB"
.I "PLOCATE_DB"

.SH DESCRIPTION
.B plocate\-build
creates an index from 
.BR plocate (1)
from an index earlier generated by
.BR updatedb (8)
from the
.BR mlocate (1)
package. Most users would rather want to use plocate's own
.BR updatedb (8),
which is more direct.
If
.I PLOCATE_DB
already exists, it will be overwritten (non-atomically). The file is
created such that it is not world-readable, as the final access check
is done by
.BR plocate (1)
during the search.

.SH OPTIONS
.TP
\fB\-b\fR, \fB\-\-block\-size\fR \fISIZE\fR
Create blocks containing
.I SIZE
filenames each, compress them together, and treat them as the same element
in the posting lists. This makes the index smaller (because the compression
algorithm gets more context to work with, and because there are fewer elements
in each posting list), but also makes posting lists less precise, moving more
work to weeding out false positives after posting list intersection.

Making this number larger will make linear search (for \fB\-\-regex\fR,
or very short patterns) faster, with diminishing returns around 256 filenames
per block. However, making it too large will cause more false positives,
reducing overall performance for typical queries. The default value is 32.
Setting it to 1 makes one (compressed) block per filename.

.TP
\fB\-p\fR, \fB\-\-plaintext\fR
Treat the input as plain text, with entries delimited by newlines,
instead of an mlocate database.

.TP
\fB\-l\fR, \fB\-\-require\-visibility\fR \fIFLAG\fR
Set the \*(lqrequire file visibility before reporting it\*(rq flag in the
generated database to \fIFLAG\fR. The default is yes, even for an mlocate
database with the flag originally set to no (although the latter may
change in the future).

.TP
.B \-\-help
Print out usage information, then exit successfully.

.TP
.B \-\-version
Print out version information, then exit successfully.

.SH AUTHOR
Steinar H. Gunderson <steinar+plocate@gunderson.no>

.SH SEE ALSO
\fBplocate\fP(1),
\fB/etc/cron.daily/plocate\fR (which is called
\fBupdate\-plocate.sh\fR in the source distribution)


NEWS
plocate 1.1.19, June 9th, 2023

  - Various bugfixes; thanks to Tim Landscheidt, Nick Alcock
    and sudhanshu goswami. In particular, this fixes a crash
    with liburing 2.4.


plocate 1.1.18, January 28th, 2023

  - Support the --require-visibility flag in plocate-build.

  - Various bugfixes.


plocate 1.1.17, November 21st, 2022

  - Return 1 if no matches were found. Patch by Clément Pit-Claudel.

  - Various portability patches, thanks to Maciej Adamczyk,
    Alfred Persson Forsberg, and the Alpine Linux packaging.


plocate 1.1.16, June 4th, 2022

  - Various bugfixes.


plocate 1.1.15, January 31st, 2022

  - Various bugfixes.


plocate 1.1.14, December 27th, 2021

  - Add an option --add-single-prunepath.
    
    This makes it possible to prune a path with spaces in the file name,
    which was previously impossible.

  - Documentation updates.


plocate 1.1.13, November 6th, 2021

  - Various bugfixes.


plocate 1.1.12, October 5th, 2021

  - Implement the -N/--literal option. This matches the GNU coreutils option
    to turn off the automatic quoting of special characters when printing to
    a tty.

  - Various bugfixes, including from Arsen Arsenović.


plocate 1.1.11, September 5th, 2021

  - Fix a crash with --existing if scanning all blocks
    (ie., very short patterns, or --regex), by turning off
    asynchronous stat() in this case. Reported by Michael Arndt.


plocate 1.1.10, August 23rd, 2021

  - Various bugfixes, including from the Void Linux repository
    and from Arsen Arsenović.


plocate 1.1.9, August 18th, 2021

  - Support the -e (--existing) option from mlocate.

  - Make it possible to set systemunitdir explicitly.
    Based on a patch by Zbigniew Jędrzejewski-Szmek.


plocate 1.1.8, June 12th, 2021

  - Various bugfixes.


plocate 1.1.7, April 7th, 2021

  - Correct randomization of updatedb systemd unit start time.
    Patch by Vladimir Lomov.


plocate 1.1.6, April 2nd, 2021

  - Support searching multiple plocate databases, including the LOCATE_PATH
    environment variable. See the plocate(1) man page for more information.

  - Fix an issue where updatedb would not recurse into directories on
    certain filesystems, in particular the deprecated XFS V4.

  - Randomize updatedb systemd unit start time. Suggested by Calum McConnell.


plocate 1.1.5, February 24th, 2021

  - Various bugfixes.


plocate 1.1.4, February 14th, 2021

  - updatedb now uses ~15% less CPU time.

  - Installs a file CACHEDIR.tag into /var/lib/plocate, to mark the directory
    as autogenerated. Suggested by Marco d'Itri.

  - Manpage fixes; patch by Jakub Wilk.


plocate 1.1.3, December 19th, 2020

  - Various portability fixes.


plocate 1.1.2, December 10th, 2020

  - Various bugfixes.


plocate 1.1.1, December 5th, 2020

  - Fix an issue where the database could be built with the wrong
    check_visibility flag.


plocate 1.1.0, December 2nd, 2020

  - Include a native updatedb, based on code and man pages from mlocate's
    updatedb (but heavily tweaked and reworked). This removes the dependency
    on mlocate's database, at the expense of 1% larger plocate database.
    plocate-build does not write the needed timestamps, so the first upgrade
    from mlocate to native plocate requires a full rescan. Subsequent rescans
    will be about as fast as updatedb.mlocate + plocate-build. plocate-build
    is now deprecated.

  - The path name now defaults to /var/lib/plocate, and the group name now
    defaults to plocate. Both can be changed at configure time.

  - Databases are now written near-atomically. An aborted updatedb or
    plocate-build will never leave a stray file.

  - Escape filenames with backticks in them.


plocate 1.0.7, October 31st, 2020

  - Fix an infinite loop when encountering invalid UTF-8 in file names.
    Bug report and patch by Leah Neukirchen.

  - Typo fixes in man pages; patch by Leah Neukirchen.

  - Portability fixes.


plocate 1.0.6, October 30th, 2020

  - Escape unprintable characters, quotes, newlines and the likes
    when outputting file names to a terminal, similar to GNU ls.
    This is to reduce the security impact of malicious filenames
    (e.g. containing color or more nefarious escape codes).
    It is not active when using -0, or when piping to another program.

  - Support building databases from plaintext files (plocate-build -p).

  - Portability fixes.


plocate 1.0.5, October 17th, 2020

  - Implement the -b/--basename option.

  - Various bugfixes.


plocate 1.0.4, October 16th, 2020

  - Fix a bug where plocate.db would be evicted from the OS cache
    (the behavior was only meant for benchmarking).

  - On fallback linear scans, such as regexes or very short patterns,
    use multithreading to speed up the search.


plocate 1.0.3, October 15th, 2020

  - Portability fixes; plocate will now compile and run on non-x86
    architectures (although without SIMD optimizations). Note that
    plocate.db is not portable across endians, but this shouldn't
    be a big problem in practice.

  - Filenames are now compressed using a shared zstd dictionary,
    which makes plocate.db ~7% smaller, and linear scans ~20% faster.
    This makes plocate-build ~20% slower, though, since it needs to
    scan through mlocate.db in a special pre-pass to sample filenames
    for the dictionary.

    This changes the dictionary format (from version 0 to version 1),
    but the new plocate can still read version 0 dictionaries,
    so the transition should be seamless.

  - Various bugfixes.


plocate 1.0.2, October 12th, 2020

  - Various bugfixes.


plocate 1.0.1, October 11th, 2020

  - Unbreak linear scanning (for very short patterns, or regexes).


plocate 1.0.0, October 11th, 2020

  - Initial release.


plocate-build.cpp
#include "database-builder.h"
#include "db.h"
#include "dprintf.h"

#include <algorithm>
#include <arpa/inet.h>
#include <assert.h>
#include <chrono>
#include <getopt.h>
#include <iosfwd>
#include <locale.h>
#include <math.h>
#include <memory>
#include <random>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <string>
#include <sys/stat.h>
#include <utility>
#include <vector>

using namespace std;
using namespace std::chrono;

bool use_debug = false;

enum {
	DBE_NORMAL = 0, /* A non-directory file */
	DBE_DIRECTORY = 1, /* A directory */
	DBE_END = 2 /* End of directory contents; contains no name */
};

// From mlocate.
struct db_header {
	uint8_t magic[8];
	uint32_t conf_size;
	uint8_t version;
	uint8_t check_visibility;
	uint8_t pad[2];
};

// From mlocate.
struct db_directory {
	uint64_t time_sec;
	uint32_t time_nsec;
	uint8_t pad[4];
};

string read_cstr(FILE *fp)
{
	string ret;
	for (;;) {
		int ch = getc(fp);
		if (ch == -1) {
			perror("getc");
			exit(1);
		}
		if (ch == 0) {
			return ret;
		}
		ret.push_back(ch);
	}
}

void handle_directory(FILE *fp, DatabaseReceiver *receiver)
{
	db_directory dummy;
	if (fread(&dummy, sizeof(dummy), 1, fp) != 1) {
		if (feof(fp)) {
			return;
		} else {
			perror("fread");
		}
	}

	string dir_path = read_cstr(fp);
	if (dir_path == "/") {
		dir_path = "";
	}

	for (;;) {
		int type = getc(fp);
		if (type == DBE_NORMAL) {
			string filename = read_cstr(fp);
			receiver->add_file(dir_path + "/" + filename, unknown_dir_time);
		} else if (type == DBE_DIRECTORY) {
			string dirname = read_cstr(fp);
			receiver->add_file(dir_path + "/" + dirname, unknown_dir_time);
		} else {
			return;  // Probably end.
		}
	}
}

void read_plaintext(FILE *fp, DatabaseReceiver *receiver)
{
	if (fseek(fp, 0, SEEK_SET) != 0) {
		perror("fseek");
		exit(1);
	}

	while (!feof(fp)) {
		char buf[1024];
		if (fgets(buf, sizeof(buf), fp) == nullptr) {
			break;
		}
		string s(buf);
		assert(!s.empty());
		while (s.back() != '\n' && !feof(fp)) {
			// The string was longer than the buffer, so read again.
			if (fgets(buf, sizeof(buf), fp) == nullptr) {
				break;
			}
			s += buf;
		}
		if (!s.empty() && s.back() == '\n')
			s.pop_back();
		receiver->add_file(move(s), unknown_dir_time);
	}
}

void read_mlocate(FILE *fp, DatabaseReceiver *receiver)
{
	if (fseek(fp, 0, SEEK_SET) != 0) {
		perror("fseek");
		exit(1);
	}

	db_header hdr;
	if (fread(&hdr, sizeof(hdr), 1, fp) != 1) {
		perror("short read");
		exit(1);
	}

	// TODO: Care about the base path.
	string path = read_cstr(fp);

	if (fseek(fp, ntohl(hdr.conf_size), SEEK_CUR) != 0) {
		perror("skip conf block");
		exit(1);
	}

	while (!feof(fp)) {
		handle_directory(fp, receiver);
	}
}

void do_build(const char *infile, const char *outfile, int block_size, bool plaintext, bool check_visibility)
{
	FILE *infp = fopen(infile, "rb");
	if (infp == nullptr) {
		perror(infile);
		exit(1);
	}

	// Train the dictionary by sampling real blocks.
	// The documentation for ZDICT_trainFromBuffer() claims that a reasonable
	// dictionary size is ~100 kB, but 1 kB seems to actually compress better for us,
	// and decompress just as fast.
	DictionaryBuilder builder(/*blocks_to_keep=*/1000, block_size);
	if (plaintext) {
		read_plaintext(infp, &builder);
	} else {
		read_mlocate(infp, &builder);
	}
	string dictionary = builder.train(1024);

	DatabaseBuilder db(outfile, /*owner=*/-1, block_size, dictionary, check_visibility);
	DatabaseReceiver *corpus = db.start_corpus(/*store_dir_times=*/false);
	if (plaintext) {
		read_plaintext(infp, corpus);
	} else {
		read_mlocate(infp, corpus);
	}
	fclose(infp);

	dprintf("Read %zu files from %s\n", corpus->num_files_seen(), infile);
	db.finish_corpus();
}

void usage()
{
	printf(
		"Usage: plocate-build MLOCATE_DB PLOCATE_DB\n"
		"\n"
		"Generate plocate index from mlocate.db, typically /var/lib/mlocate/mlocate.db.\n"
		"Normally, the destination should be /var/lib/mlocate/plocate.db.\n"
		"\n"
		"  -b, --block-size SIZE  number of filenames to store in each block (default 32)\n"
		"  -p, --plaintext        input is a plaintext file, not an mlocate database\n"
	        "  -l, --require-visibility FLAG  check visibility before reporting files\n"
		"      --help             print this help\n"
		"      --version          print version information\n");
}

void version()
{
	printf("plocate-build %s\n", PACKAGE_VERSION);
	printf("Copyright 2020 Steinar H. Gunderson\n");
	printf("License GPLv2+: GNU GPL version 2 or later <https://gnu.org/licenses/gpl.html>.\n");
	printf("This is free software: you are free to change and redistribute it.\n");
	printf("There is NO WARRANTY, to the extent permitted by law.\n");
}

bool parse_bool(const string &str, bool *result)
{
	if (str == "0" || str == "no") {
		*result = false;
		return true;
	}
	if (str == "1" || str == "yes") {
		*result = true;
		return true;
	}
	return false;
}

int main(int argc, char **argv)
{
	static const struct option long_options[] = {
		{ "block-size", required_argument, 0, 'b' },
		{ "plaintext", no_argument, 0, 'p' },
		{ "require-visibility", required_argument, 0, 'l' },
		{ "help", no_argument, 0, 'h' },
		{ "version", no_argument, 0, 'V' },
		{ "debug", no_argument, 0, 'D' },  // Not documented.
		{ 0, 0, 0, 0 }
	};

	int block_size = 32;
	bool plaintext = false;
	bool check_visibility = true;

	setlocale(LC_ALL, "");
	for (;;) {
		int option_index = 0;
		int c = getopt_long(argc, argv, "b:hpl:VD", long_options, &option_index);
		if (c == -1) {
			break;
		}
		switch (c) {
		case 'b':
			block_size = atoi(optarg);
			break;
		case 'p':
			plaintext = true;
			break;
		case 'l':
			if (!parse_bool(optarg, &check_visibility) != 0) {
				fprintf(stderr, "plocate-build: invalid value `%s' for --%s\n",
					 optarg, "require-visibility");
				exit(EXIT_FAILURE);
			}
			break;
		case 'h':
			usage();
			exit(0);
		case 'V':
			version();
			exit(0);
		case 'D':
			use_debug = true;
			break;
		default:
			exit(1);
		}
	}

	if (argc - optind != 2) {
		usage();
		exit(1);
	}

	do_build(argv[optind], argv[optind + 1], block_size, plaintext, check_visibility);
	exit(EXIT_SUCCESS);
}
